### PR TITLE
Use an ingestion timestamp to detect whether the search/pulldata handler should re-process new data file

### LINKED ISSUE #3335: The search()/pulldata() handler consumes files meant to be used as external secondary instances
To reproduce this issue, #3334 first needs to be addressed.

Then, try a form like [this one](https://docs.google.com/spreadsheets/d/1xZDS2XqwbMcgqa5nvi5_Ug03WkijogcQh4yPNexqaE8/edit#gid=0) that attempts to use a CSV file as an external secondary instance. It works the first time the form is opened but fails the second time because the CSV has been imported to a db by the search()/pulldata() machinery and renamed.

Some options I see to address this:
- instead of renaming CSVs to mark them as ingested, leave the files untouched and use a file hash to detect a change
- instead of greedily putting all CSVs in a DB, only read in the ones used for search()/pulldata() (or not used as external secondary instance)

### PR DESCRIPTION
**Pull request by lognaturel**
Closes #3335

I cherry picked @OpenDataNerd's excellent commits from https://github.com/OpenDataNerd/collect/tree/csv_external_data. The XLSForm documentation has been updated at http://xlsform.org/en/#multiple-choice-from-file to recommend using CSV external instances so it's now feeling more urgent to get this fixed.

#### What has been done to verify that this works as intended?

Ran automated tests. Opened a form with external secondary CSV instance multiple times and one with search() multiple times. I also tried a form with an `itemsets.csv` file even though there should be no interaction with that feature.

#### Why is this the best possible solution? Were any other approaches considered?

The issue discusses alternatives. In particular, I had suggested using a file hash to identify changes. I think the time stamp solution is simpler and just as effective. We also could have considered storing the time stamps in an auxiliary file or something like that but adding a metadata table keeps things well-organized. 

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Regression risk is around everything having to do with external data. 
#### Do we need any specific form for testing your changes? If so, please attach one.
Config for a Central project with two forms: one with a CSV external instance and one with a search() CSV:
csv

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No. 
#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by seadowg** [comment]
I think in the interest of getting this through it would be good for @opendatakit/testers to look at this in parallel to review so I'll mark this as "needs testing" now

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import org.odk.collect.android.utilities.CustomSQLiteQueryBuilder;
+import org.odk.collect.android.utilities.FileUtils;
+import org.odk.collect.android.utilities.SQLiteUtils;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ExternalDataTest {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: Is this really `ExternalDataReaderTest`?

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    @Before
+    public void setUp() {
+        File formPath = new File(new StoragePathProvider().getDirPath(StorageSubdirectory.FORMS) + File.separator + SIMPLE_SEARCH_EXTERNAL_CSV_FORM_FILENAME);
+        File mediaDir = FileUtils.getFormMediaDir(formPath);
+        csvFile = new File(mediaDir + File.separator + SIMPLE_SEARCH_EXTERNAL_CSV_FILENAME);
+        dbFile = new File(mediaDir + File.separator + SIMPLE_SEARCH_EXTERNAL_DB_FILENAME);
+    }
+
+    @Rule
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: Could this run in Robolectric instead?

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            .around(new CopyFormRule(SIMPLE_SEARCH_EXTERNAL_CSV_FORM_FILENAME, Collections.singletonList(SIMPLE_SEARCH_EXTERNAL_CSV_FILENAME)));
+
+    private static Map<String, File> makeExternalDataMap(File csvFile) {
+        Map<String, File> externalDataMap = new HashMap<>();
+        externalDataMap.put(SIMPLE_SEARCH_EXTERNAL_CSV_NAME, csvFile);
+        return externalDataMap;
+    }
+
+    @Test
+    public void testCreateDBonImportCSV() {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: These tests are pretty dense and I'm having a hard time understanding what they are driving at. I think using a different style for the names would be helpful. This test could be `doImport_createsDataAndMetaDatabaseTables`. That way we know what the subject is (`ExternalDataReaderImpl#doImport`) and what our expectation is pretty clearly.

File: collect_app/src/androidTest/java/org/odk/collect/android/tasks/FormLoaderTaskTest.java [hunk 4]
```diff
@@ -14,13 +15,19 @@
 
 import java.io.File;
 import java.util.Arrays;
+import java.util.Collections;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.notNullValue;
 
 public class FormLoaderTaskTest {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/tasks/FormLoaderTaskTest.java:23]: I'd love to say "this should be Robolectric" but I think that would involve picking the task apart. The `XTaskTest` classes are pretty much the prime candidates for moving out of Espresso so I'm confident that will happen later.

**Review by seadowg — CHANGES_REQUESTED** [review]
I found some of the tests hard to follow so I've left comments about that for ways that could hopefully be improved.

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java (hunk 1 above)
Comment (lognaturel) [reply collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: Yes, that sounds right. Will change!

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java (hunk 2 above)
Comment (lognaturel) [reply collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: I think so, good call. Let me explore that...

File: collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java (hunk 3 above)
Comment (lognaturel) [reply collect_app/src/androidTest/java/org/odk/collect/android/external/ExternalDataTest.java]: Yes, sounds good.

File: collect_app/src/androidTest/assets/forms/simple-search-external-csv.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lognaturel) [inline collect_app/src/androidTest/assets/forms/simple-search-external-csv.xml:1]: It's too bad to have duplicates of these test files in `androidTest` and `test`. I think it's the right thing to do the Robolectric tests with real form files, though.

This might actually be a good reason for all tests to live in the same source set, as the androidx libs want.

File: collect_app/src/androidTest/java/org/odk/collect/android/tasks/FormLoaderTaskTest.java (hunk 4 above)
Comment (lognaturel) [reply collect_app/src/androidTest/java/org/odk/collect/android/tasks/FormLoaderTaskTest.java:23]: Yeah, agreed that coming back to this sounds good. 

I think the new Robolectric tests may actually offer sufficient coverage. There is something comforting about knowing that these are even more end-to-end, though.

File: collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    private static File dbFile;
+
+    private static Map<String, File> formDefToCsvMedia;
+
+    @Before
+    public void setUp() throws IOException {
+        ShadowEnvironment.setExternalStorageState(Environment.MEDIA_MOUNTED);
+        new StorageInitializer().createOdkDirsOnStorage();
+
+        File formFile = new File(new StoragePathProvider().getDirPath(StorageSubdirectory.FORMS) + File.separator + SIMPLE_SEARCH_EXTERNAL_CSV_FORM_FILENAME);
```
Comment (lognaturel) [inline collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java:59]: This work here is basically exactly what `CopyFormRule` and the methods in `FormLoadingUtils` do but they're in different sourcesets. Again, maybe a reason to have all tests in one. Or if there is an elegant way to share that code, please feel free to modify, @seadowg.

File: collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+        db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        assertThat("metadata table should be recreated", SQLiteUtils.doesTableExist(db, EXTERNAL_METADATA_TABLE_NAME));
+        db.close();
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfFileIsUpdated() throws IOException, InterruptedException {
```
Comment (lognaturel) [inline collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java:161]: Note that this test has changed. It previously faked an update by changing the db timestamp. Now it actually tests a real file change.

File: collect_app/src/androidTest/assets/forms/simple-search-external-csv.xml (hunk 5 above)
Comment (seadowg) [reply collect_app/src/androidTest/assets/forms/simple-search-external-csv.xml:1]: I think it's fine to duplicate. A nice solution is to bundle the test forms in gradle submodule if it gets annoying. I have thoughts on everything living in the same set but they can be a fun discussion another time!

File: collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java (hunk 6 above)
Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java:59]: I'm happy with this for now!

File: collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        assertThat(dbFile.exists(), is(true));
+
+        SQLiteDatabase db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        assertThat(db.rawQuery(SELECT_ALL_DATA_QUERY, null).getCount(), is(3));
+
+        long originalTimestamp = csvFile.lastModified();
+        long metadataLastModified = ExternalSQLiteOpenHelper.getLastImportTimestamp(db, EXTERNAL_METADATA_TABLE_NAME, csvFile);
+        assertThat(metadataLastModified, is(originalTimestamp));
+
+        Thread.sleep(1000); // this test is so fast that we need to guarantee the modified time changes
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java:174]: I think there is a way to fake this with fake file layer. Can't remember the name of the framework though. We can stick with this right now as I'm pretty sure it's still deterministic.

**Review by seadowg — APPROVED** [review]

**Comment by kkrawczyk123** [comment]
@lognaturel I have identified one weird scenario using that form as it is visible on master too I reported it separately in #3682. 
The pr was verified on Androids: 5.1, 6.0, 7.0, 8.1, 9.0 and 10.
I have cooperated with @mmarciniak90 and we have verified cases:
- download attached forms from Central
- fill forms
- edit filled forms
- send finalized forms
- migrate and again fill, edit, send both forms
- download the same forms from Google drive: fill, edit, send, migrate, again fill, edit
- download form with external CSV and last save: fill, edit, send, migrate, again fill, edit
- check other forms with external CSV files: fill, edit, send, migrate, again fill, edit

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
